File: 2015/Go/3/part2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Point struct {
	X int
	Y int
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	houses := make(map[Point]int)

	sx, sy := 0, 0
	rx, ry := 0, 0

	houses[Point{0, 0}] = 2

	for scanner.Scan() {
		line := scanner.Text()
		for i, char := range line {
			switch char {
			case '^':
				if i%2 == 0 {
					sy++
					houses[Point{sx, sy}]++
				} else if i%2 != 0 {
					ry++
					houses[Point{rx, ry}]++
				}
			case 'v':
				if i%2 == 0 {
					sy--
					houses[Point{sx, sy}]++
				} else if i%2 != 0 {
					ry--
					houses[Point{rx, ry}]++
				}
			case '>':
				if i%2 == 0 {
					sx++
					houses[Point{sx, sy}]++
				} else if i%2 != 0 {
					rx++
					houses[Point{rx, ry}]++
				}
			case '<':
				if i%2 == 0 {
					sx--
					houses[Point{sx, sy}]++
				} else if i%2 != 0 {
					rx--
					houses[Point{rx, ry}]++
				}
			default:
				panic("Character not assept.")
			}
		}
		fmt.Printf("How many houses receive at least one present? (%d)\n", len(houses))
	}
}

File: 2015/Rust/3/part2/src/main.rs
use std::collections::HashMap;

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
struct Point {
    x: i32,
    y: i32,
}

fn main() {
    let raw_text = include_str!("input.txt");

    let mut houses: HashMap<Point, i32> = HashMap::new();

    let point = Point { x: 0, y: 0 };

    let mut _sx = point.x;
    let mut _sy = point.y;

    let mut _rx = point.x;
    let mut _ry = point.y;

    houses.insert(Point { x: 0, y: 0 }, 2);

    for (i, c) in raw_text.chars().enumerate() {
        let (x_current, y_current) = if i % 2 == 0 {
            (&mut _sx, &mut _sy)
        } else {
            (&mut _rx, &mut _ry)
        };
        match c {
            '^' => *y_current += 1,
            'v' => *y_current -= 1,
            '>' => *x_current += 1,
            '<' => *x_current -= 1,
            _ => println!("This character is not acept."),
        }
        houses.insert(
            Point {
                x: *x_current,
                y: *y_current,
            },
            1,
        );
    }
    println!(
        "How many houses receive at least one present? {}",
        houses.len()
    );
}
